### ISSUE TITLE
Downloads never start

### ISSUE THREAD

**Issue by petercockroach** [deepjyoti30/ytmdl-web-v2#8]
# Bug report

## Checklist

- [x] I have searched for similar bugs in the issues.
- [x] I will add the bug label to this issue.

## How was Ytmdl Web run

- [x] Locally
- [x] On the official page

## Browser Details

- Name: Safari and Brave
- Version: Safari Version 14.1 (16611.1.18.1.2)

## Console Log

No obvious errors

## Description

Serach for anything. When clicking the result, it either goes to "Getting the song ready for you" or it takes me to another page of apparent search results which seem unrelated. Occurs locally and on official page.

## Expected Behaviour

Download starts.

**Comment by deepjyoti30** [comment]
@petercockroach That is the expected behaviour!? Once you click the search result it will work on the song and after that show you the download button.

Are you trying to say once the metadata is selected, it goes into and unending loop? 

I just tested the official app and it worked totally well.!

**Comment by petercockroach** [comment]
I get the "Getting the song ready for you" animation, then nothing happens (currently testing this on your official site). Yesterday, I was getting prompted with new unrelated search results but at no point does the download ever start.

**Comment by deepjyoti30** [comment]
@petercockroach I did test the official site the last I commented to you. I think there might be some issue but that's probably because the server is not able to handle so much load. I have currently deployed the API on heroku (had some credits and since my own server was blocked by YouTube) and it does work.

**Comment by petercockroach** [comment]
FWIW, the issue occurs locally on my as well so the cause is something common between the official site and local deployment.

**Comment by deepjyoti30** [comment]
@petercockroach Yeah, the issue is not from the frontend. If it is occuring, it is from the backend.

However, it is very weird because everytime I use the app and track the logs, I see a proper response and everything works fine.

As I suspected, this might be happening due to too much requests on the server.

Also is the issue happening constantly with you? In all browsers as well?

**Comment by petercockroach** [comment]
Occurs in Brave (Chrome) and Safari, and yes, constantly.

**Comment by deepjyoti30** [comment]
@petercockroach It is very weird. Can you consider trying to download in an incognito tab perhaps? That will help to know if the issue is because of some kind of local issue.

I just tried the app as I am writing this reply and it worked totally fine, no issues at all.

**Comment by petercockroach** [comment]
I reproduced it on an entirely different device using cellular data which means it’s not the browser/cache, it’s not my OS, and it’s not my network

**Comment by boxdox** [comment]
Hey @deepjyoti30, this seems to be CORS issue for me. You might wanna check your heroku app.

```
Access to fetch at 'https://ytmdlapi.herokuapp.com/download' from origin 'https://ytmdl.deepjyoti30.dev' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Comment by deepjyoti30** [comment]
@boxdox Just took a look at the CORS issue. It is happening because the Heroku requests are timing out due to excessive memory use.

Thus when the API tries to download the song in the background, it is failing.

Regarding the service worker registration, I was not aware of the issue because I was loading the app in my browser which had the worker installed from before but I took a look at the app on Incognito and saw the issue.

**Comment by boxdox** [comment]
> @boxdox Just took a look at the CORS issue. It is happening because the Heroku requests are timing out due to excessive memory use.
> 
> Thus when the API tries to download the song in the background, it is failing.
> 
> Regarding the service worker registration, I was not aware of the issue because I was loading the app in my browser which had the worker installed from before but I took a look at the app on Incognito and saw the issue.

Yes, the new service worker was broken for new users. Maybe you should host this on a VPS, if you get some funding.

**Comment by deepjyoti30** [comment]
@boxdox  I am already hosting the webapp on a VPS.

Both the API and the webapp were on my server. However, after the launch the API got so many hits that my severs IP was blocked by YouTube. This is why I moved to Heroku in order to fix the issue ASAP.

**Comment by boxdox** [comment]
I can give you some suggestions:
1. Add a rotating proxy (still you have to beef up your servers for more clients)
2. Instead of downloading the video on the backend, just extract the download URL and send it to client for downloading (I have not tested if there will be any IP issues as download URL is based on IP), then you can use something like ffmpeg-wasm to convert it to audio on frontend.
3. Compile the whole Youtube-dl with ffmpeg into a single wasm (slow for first time, subsequent visits will be faster)

**Comment by deepjyoti30** [comment]
@boxdox Thanks for the suggestions.

1. Already considered this. Thought of installing a load balancer but that would obviously require more servers. When I was using my personal server to serve, I was using a 10$/month plan so 2/3 more servers would be twice/thrice of that. That is completely out of scope. The project doesn't earn in any way except donations. The donations (since release) won't even make up for one server for a month. Thus can't go with this option.

   A lot of people suggested adding ads and I am currently considering that. I just want to make the app self sustainable, so that all the revenue is able to pay the server costs. That way I can consider investing more on the infrastructure.
   
2. The whole idea of the webapp was to build something that works like `ytmdl`. Thus I just built the API on top of `ytmdl`. So the song is downloaded first and then the metadata is added to the file and then it is served.

3. Won't this be similar to moving away from the actual ytmdl? Besides that, it would be like reinventing the wheel to build an webapp with all the features that the `CLI` version supports.

**Comment by boxdox** [comment]
> @boxdox Thanks for the suggestions.
> 
> 1. Already considered this. Thought of installing a load balancer but that would obviously require more servers. When I was using my personal server to serve, I was using a 10$/month plan so 2/3 more servers would be twice/thrice of that. That is completely out of scope. The project doesn't earn in any way except donations. The donations (since release) won't even make up for one server for a month. Thus can't go with this option.
> 
>    A lot of people suggested adding ads and I am currently considering that. I just want to make the app self sustainable, so that all the revenue is able to pay the server costs. That way I can consider investing more on the infrastructure.
>    
> 2. The whole idea of the webapp was to build something that works like `ytmdl`. Thus I just built the API on top of `ytmdl`. So the song is downloaded first and then the metadata is added to the file and then it is served.
> 
> 3. Won't this be similar to moving away from the actual ytmdl? Besides that, it would be like reinventing the wheel to build an webapp with all the features that the `CLI` version supports.

Yeah makes sense, I didn't consider you have an existing API. So your only option is getting some sponsorship. Ads will still not cover the costs anytime soon. Maybe you can work on some user plans and limit guest users?

**Comment by deepjyoti30** [comment]
@boxdox I was considering that as an option too. For example, users that would pay a certain fee will get a certain number of allowed downloads in a month.

However, this would require some user login stuff and I totally don't want to force anyone to signup etc. I will though consider this.

I think it can be a valid plan! Only if users would sign up tho! :sweat_smile:

**Comment by boxdox** [comment]
@deepjyoti30 you can have a guest tier, a free user tier and a paid tier. That way you don't force someone to signup, but if they like your service, they can register or maybe buy a plan.
